File: CPP/BOJ-S4-1940-Joo_mong.cpp
#include <bits/stdc++.h>
using namespace std;

int n, m, a[15004], cnt;

int main()
{
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  freopen("input.txt", "r", stdin);
  freopen("output.txt", "w", stdout);

  cin >> n >> m;

  for (int i = 0; i < n; ++i)
  {
    cin >> a[i];
  }

  // 고유번호 100,000이면 두 개 조합 시 최대 200,000.
  // m은 최대 10,000,000으로, 200,000 초과하는 경우 탐색 불필요.
  if (m > 200000)
    cout << 0 << '\n';
  else
  {
    for (int i = 0; i < n; ++i)
    {
      for (int j = i + 1; j < n; ++j)
      {
        if (a[i] + a[j] == m)
          ++cnt;
      }
    }
    cout << cnt << '\n';
  }

  return 0;
}

File: JS/BOJ-S4-1940-Joo_mong.js
'use strict';

const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../input.txt';
let input = fs
  .readFileSync(filePath)
  .toString()
  .trim()
  .split('\n')
  .map((v) => v.trimEnd());

let output = [];
const solution = (input) => {
  const [n, m] = [+input[0], +input[1]];
  let nums = input[2]
    .split(' ')
    .map((v) => +v)
    .sort((a, b) => a - b);

  let cnt = 0;
  const comb = (idx = 0, arr = []) => {
    if (arr.length == 2) {
      if (arr[0] + arr[1] == m) {
        ++cnt;
      }
      return;
    }
    for (let i = idx; i < n; ++i) {
      arr.push(nums[i]);
      comb(i + 1, arr);
      arr.pop();
    }
  };

  comb();

  output.push(cnt);
};

solution(input);
fs.writeFileSync('/dev/stdout', output.join('').trimEnd());
// fs.writeFileSync(
//   '../output.txt',
//   output.join('').trimEnd(),
//   (err) => console.error
// );
